src/Utils.cs

```diff
@@ -1,44 +1,51 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
 
 namespace SpacetimeDB
 {
-    public static class Utils
+    internal readonly struct ByteArrayComparer : IEqualityComparer<byte[]>
     {
-        public static bool ByteArrayCompare(byte[] a1, byte[] a2)
+        public static readonly ByteArrayComparer Instance = new();
+
+        public bool Equals(byte[]? left, byte[]? right)
         {
-            if (a1 == null || a2 == null)
-                return a1 == a2;
+            if (ReferenceEquals(left, right))
+            {
+                return true;
+            }
 
-            if (a1.Length != a2.Length)
+            if (left is null || right is null || left.Length != right.Length)
+            {
                 return false;
+            }
 
-            for (int i = 0; i < a1.Length; i++)
-                if (a1[i] != a2[i])
-                    return false;
+            return EqualsUnvectorized(left, right);
 
-            return true;
         }
-    }
 
-    public class ByteArrayComparer : IEqualityComparer<byte[]>
-    {
-        public bool Equals(byte[] x, byte[] y)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsUnvectorized(byte[] left, byte[] right)
         {
-            return Utils.ByteArrayCompare(x, y);
+            for (int i = 0; i < left.Length; i++)
+            {
+                if (left[i] != right[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
         }
 
         public int GetHashCode(byte[] obj)
         {
-            if (obj == null)
-                return 0;
-            int sum = 0;
-            for (int i = 0; i < obj.Length; i++)
-                sum += obj[i];
-            return sum;
+            int hash = 17;
+            foreach (byte b in obj)
+            {
+                hash = hash * 31 + b;
+            }
+            return hash;
```

Comment:
We could think about pulling in a fast hash as a dependency if needed, but I think this should work fine.